### ISSUE TITLE
Misleading compiler error when calling overloaded function with incorrect params

### ISSUE THREAD

**Issue by UselessPickles** [Microsoft/TypeScript#20697]
**TypeScript Version:**  2.6.2

**Code**
(with all strict compiler options enabled)
```ts
declare function foo(value: number, other: number): void;
declare function foo(value: number | null, other: string): void;

declare const x: number;

foo(x, true);
```

**Expected behavior:**
A compiler error complaining that the second arg (type `boolean`) is not assignable to type `number`, using the first overloaded signature.

I would expect the compiler error to be based on the "most applicable" overload signature. If multiple signatures are equally "applicable", then the order of the signatures should be the tie breaker.

**Actual behavior:**
A compiler error complaining that the second arg cannot be assigned to type `string`, using the last overloaded signature. When none of the overload signatures are valid, the compiler seems to always use the last overload signature for creating the error message. This can be very misleading.

![screen shot 2017-12-14 at 11 40 46 am](https://user-images.githubusercontent.com/22800095/34003723-01aac9a6-e0c4-11e7-8528-f01de6eeaa8e.png)

In this situation, it is more likely that the developer should be guided toward providing a `number` as the second arg, because the type of their `x` variable does not require the second overload signature to be used.

**Comment by RyanCavanaugh** [comment]
See https://github.com/Microsoft/TypeScript/issues/20139#issuecomment-345788338 and many other issues suggesting that we pick the "best" overload

**Comment by UselessPickles** [comment]
Can't you just program the compiler to read my mind? :)

I'm sure there's many layers of complexity I'm not aware of, but this is my initial thought on selecting the "best" overload for errors:

* Pick the overload with the highest number of params that are assignable (maybe include +1 in the count if return value is assignable?).
* Use the order of the overrides as a tie-breaker (pick the first overload with the highest number of assignable params).
